src/libraries/System.IO.Pipes.AccessControl/src/System.IO.Pipes.AccessControl.csproj

```diff
@@ -1,26 +1,59 @@
+    <Compile Include="$(LibrariesProjectRoot)System.IO.Pipes\src\System\IO\Pipes\PipeSecurity.cs" />
+    <Compile Include="$(LibrariesProjectRoot)System.IO.Pipes\src\System\IO\Pipes\PipeAccessRights.cs" />
+    <Compile Include="$(LibrariesProjectRoot)System.IO.Pipes\src\System\IO\Pipes\PipeAccessRule.cs" />
+    <Compile Include="$(LibrariesProjectRoot)System.IO.Pipes\src\System\IO\Pipes\PipeAuditRule.cs" />
+    <Compile Include="$(LibrariesProjectRoot)System.IO.Pipes\src\System\IO\Pipes\PipesAclExtensions.cs" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETStandard'">
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Security.AccessControl\src\System.Security.AccessControl.csproj" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Security.Principal.Windows\src\System.Security.Principal.Windows.csproj" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
     <Reference Include="System.Runtime" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Security.AccessControl\src\System.Security.AccessControl.csproj" />
+    <Reference Include="System.IO.Pipes" Condition="'$(TargetsWindows)' != 'true' or '$(TargetPlatformIdentifier)' == ''" />
   </ItemGroup>
-  <!-- Include src projects during restore as TargetsWindows isn't set. -->
-  <ItemGroup Condition="'$(TargetsWindows)' == 'true' or '$(MSBuildRestoreSessionId)' != ''">
-    <ProjectReference Include="$(LibrariesProjectRoot)System.IO.Pipes\src\System.IO.Pipes.csproj" SkipUseReferenceAssembly="true" />
-    <ProjectReference Include="$(LibrariesProjectRoot)System.Security.AccessControl\src\System.Security.AccessControl.csproj" SkipUseReferenceAssembly="true" />
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.1'">
+    <!-- THIS IS THE WORST CODE I'VE WRITTEN IN A WHILE. We need to download the 3.1 microsoft.net.core.app.runtime pack for windows
+         as this library depends on an API only exposed in the System.IO.Pipes runtime assembly.
+         We need to come up with a better solution as this adds 30MB to a user's download and breaks sourcebuild. -->
+    <PackageDownload Include="Microsoft.NETCore.App.Runtime.win-x64" Version="[3.1.14]" />
+    <ReferencePath Include="$(NuGetPackageRoot)microsoft.netcore.app.runtime.win-x64\3.1.14\runtimes\win-x64\lib\netcoreapp3.1\System.IO.Pipes.dll" Condition="'$(TargetsWindows)' == 'true'" />
+
+    <Reference Include="System.Resources.ResourceManager" />
+  </ItemGroup>
+
+  <ItemGroup Condition="$([MSBuild]::GetTargetFrameworkVersion('$(TargetFramework)')) == '5.0'">
+    <!-- THIS IS THE WORST CODE I'VE WRITTEN IN A WHILE. We need to download the 5.0 microsoft.net.core.app.runtime pack for windows
+         as this library depends on an API only exposed in the System.IO.Pipes runtime assembly.
+         We need to come up with a better solution as this adds 30MB to a user's download and breaks sourcebuild. -->
```

Comment:
We could create a special reference assemblies for System.IO.Pipes, only for use during package build, built from source making sure they get the correct assembly version.  I think all they need are the type-defs and assembly versions so that we can produce the right typeforwards for those frameworks.

Alternatively we convince ourselves we are willing to expose https://github.com/dotnet/runtime/blob/c9c954be788c4d4d13af326d83ba206a7f2ce45d/src/libraries/NetCoreAppLibrary.props#L151-L153
And then dead-end this package, telling folks to use API exposed from ref-pack instead.

Reply:
> We could create a special reference assemblies for System.IO.Pipes

Right. That's yet another alternative which I considered but didn't write down. My only concern about this approach is that it's the only place in our graph where a lib exists just to satisfy a dependent lib as it wouldn't ship at all. We would need to carry them along (via tfm upgrades) as long as the package exists.

Another alternative would be to manually add the problematic type forwards (to System.IO.Pipes) and exclude them from being auto-generated via GenFacades. That would avoid the non-shipping System.IO.Pipes reference configurations.

Reply:
Here's what I was thinking about, could probably be cleaned up more: https://github.com/ericstj/runtime/commit/f2dd6250ce889d479707740f6572a55f9745db4e

By manually add type-forwards were you thinking compile from IL?  I don't think C# will let you forward to a type unless that type exists in a reference assembly.

Reply:
You are right, I just tried type forwarding in C# directly and the destination type must be reachable by the compiler. Thanks for the commit, will take a look.